# Date Night Movie

#### Grading:


- Code: 90 pts
- Markdown Documentation: 10 pts


In this assignment we are going to use pandas to figure out - What's the best **date-night movie**?

This assignment is going to use
- Joining
- Groupby
- Sorting


In [1]:
import os
import pandas as pd


##### Read in the movie data: `pd.read_table`

In [2]:
def get_movie_data():
    
    unames = ['user_id','gender','age','occupation','zip']
    users = pd.read_table(r'../data/users.dat',encoding="unicode_escape", sep='::', header=None, names=unames)
    
    rnames = ['user_id', 'movie_id', 'rating', 'timestamp'] 
    ratings = pd.read_table(r'../data/ratings.dat',encoding="unicode_escape", sep='::', header=None, names=rnames)
    
    mnames = ['movie_id', 'title','genres']
    movies = pd.read_table(r'../data/movies.dat',encoding="unicode_escape", sep='::', header=None, names=mnames)

    return users, ratings, movies
# implemented encoding="unicode_escape" to read the files

In [3]:
users, ratings, movies = get_movie_data()

/Users/vijaykrishna/opt/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py:311: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return func(*args, **kwargs)


In [4]:
print (users.head())
print(users.shape)

   user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455
(6040, 5)


In [5]:
print (ratings.head())
print(ratings.shape)

   user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291
(1000209, 4)


In [6]:
print (movies.head())
print(movies.shape)

   movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy
(3883, 3)


##### Clean up the `movies`

- Get the `year`
- Shorten the `title`


In [7]:
tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')
tmp.apply(lambda x:x[0] if len(x) > 0 else None)
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)

0    Toy Story
1         1995
dtype: object

In [8]:
movies['year'] = tmp[1]
movies['short_title'] = tmp[0]

In [9]:
print (movies.head())

   movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II  


##### Join the tables with `pd.merge` (20 pts)

- Merging all the given dataframes using pd.merge

In [10]:
ratings_users = pd.merge(users,ratings, left_on='user_id', right_on='user_id')
print(ratings_users)
# inner merge all of them

         user_id gender  age  occupation    zip  movie_id  rating  timestamp
0              1      F    1          10  48067      1193       5  978300760
1              1      F    1          10  48067       661       3  978302109
2              1      F    1          10  48067       914       3  978301968
3              1      F    1          10  48067      3408       4  978300275
4              1      F    1          10  48067      2355       5  978824291
...          ...    ...  ...         ...    ...       ...     ...        ...
1000204     6040      M   25           6  11106      1091       1  956716541
1000205     6040      M   25           6  11106      1094       5  956704887
1000206     6040      M   25           6  11106       562       5  956704746
1000207     6040      M   25           6  11106      1096       4  956715648
1000208     6040      M   25           6  11106      1097       4  956715569

[1000209 rows x 8 columns]


In [11]:
ratings_users_movies = pd.merge(movies,ratings_users, left_on='movie_id' , right_on = 'movie_id')
ratings_users_movies

movie_id                  title                       genres  year  \
0               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
1               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
2               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
3               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
4               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
...           ...                    ...                          ...   ...   
1000204      3952  Contender, The (2000)               Drama|Thriller  2000   
1000205      3952  Contender, The (2000)               Drama|Thriller  2000   
1000206      3952  Contender, The (2000)               Drama|Thriller  2000   
1000207      3952  Contender, The (2000)               Drama|Thriller  2000   
1000208      3952  Contender, The (2000)               Drama|Thriller  2000   

            short_title  user_id gender  age  occupation    zip  rating  \
0             Toy Story        1      F    1          10  48067       5   
1             Toy Story        6      F   50           9  55117       4   
2             Toy Story        8      M   25          12  11413       4   
3             Toy Story        9      M   25          17  61614       5   
4             Toy Story       10      F   35           1  95370       5   
...                 ...      ...    ...  ...         ...    ...     ...   
1000204  Contender, The     5812      F   25           7  92120       4   
1000205  Contender, The     5831      M   25           1  92120       3   
1000206  Contender, The     5837      M   25           7  60607       4   
1000207  Contender, The     5927      M   35          14  10003       1   
1000208  Contender, The     5998      M   18           4  61820       4   

          timestamp  
0         978824268  
1         978237008  
2         978233496  
3         978225952  
4         978226474  
...             ...  
1000204   992072099  
1000205   986223125  
1000206  1011902656  
1000207   979852537  
1000208  1001781044  

[1000209 rows x 12 columns]

##### What's the highest rated movie? (20 pts))

- From the merged dataframe extracting the total number of titles and number of occurrences in the given data

-https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html

In [12]:
counts = ratings_users_movies['title'].value_counts()
counts

American Beauty (1999)                                      3428
Star Wars: Episode IV - A New Hope (1977)                   2991
Star Wars: Episode V - The Empire Strikes Back (1980)       2990
Star Wars: Episode VI - Return of the Jedi (1983)           2883
Jurassic Park (1993)                                        2672
                                                            ... 
Kestrel's Eye (Falkens öga) (1998)                             1
Last of the High Kings, The (a.k.a. Summer Fling) (1996)       1
Condition Red (1995)                                           1
Beauty (1998)                                                  1
Soft Toilet Seats (1999)                                       1
Name: title, Length: 3706, dtype: int64

#### From the dataframe only extracting the titles with more number of ratings i.e.. atleast 500 ratings and above


- below code  picks the titles with more than 500 count and then prints our criteria based on the index in row-wise
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html

In [13]:
ratings_users_movies_1 = ratings_users_movies[ratings_users_movies['title'].isin(counts[counts > 500].index)]
ratings_users_movies_1

movie_id                    title                       genres  year  \
0              1         Toy Story (1995)  Animation|Children's|Comedy  1995   
1              1         Toy Story (1995)  Animation|Children's|Comedy  1995   
2              1         Toy Story (1995)  Animation|Children's|Comedy  1995   
3              1         Toy Story (1995)  Animation|Children's|Comedy  1995   
4              1         Toy Story (1995)  Animation|Children's|Comedy  1995   
...          ...                      ...                          ...   ...   
999418      3948  Meet the Parents (2000)                       Comedy  2000   
999419      3948  Meet the Parents (2000)                       Comedy  2000   
999420      3948  Meet the Parents (2000)                       Comedy  2000   
999421      3948  Meet the Parents (2000)                       Comedy  2000   
999422      3948  Meet the Parents (2000)                       Comedy  2000   

             short_title  user_id gender  age  occupation    zip  rating  \
0              Toy Story        1      F    1          10  48067       5   
1              Toy Story        6      F   50           9  55117       4   
2              Toy Story        8      M   25          12  11413       4   
3              Toy Story        9      M   25          17  61614       5   
4              Toy Story       10      F   35           1  95370       5   
...                  ...      ...    ...  ...         ...    ...     ...   
999418  Meet the Parents     5938      M   25           1  35401       4   
999419  Meet the Parents     5949      M   18          17  47901       2   
999420  Meet the Parents     5950      M   25           4  19713       4   
999421  Meet the Parents     5972      F   25          20  55428       2   
999422  Meet the Parents     5996      F   25           0  87114       3   

         timestamp  
0        978824268  
1        978237008  
2        978233496  
3        978225952  
4        978226474  
...            ...  
999418  1002297887  
999419   985052813  
999420  1046369637  
999421   986916093  
999422   972309420  

[594212 rows x 12 columns]

##### In order to get the top ratings I implement a mean to get the mean values of all the user ratings. Implemented a groupby function on 'title' column for the mean values of the 'ratings' column
###### reference :
- https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.core.groupby.DataFrameGroupBy.agg.html

In [14]:
ratings_users_movies_mean = ratings_users_movies_1.groupby('title')['rating'].agg(['mean'])
ratings_users_movies_mean

mean
title                                          
10 Things I Hate About You (1999)      3.422857
101 Dalmatians (1961)                  3.596460
12 Angry Men (1957)                    4.295455
13th Warrior, The (1999)               3.158667
20,000 Leagues Under the Sea (1954)    3.702609
...                                         ...
X-Files: Fight the Future, The (1998)  3.492972
X-Men (2000)                           3.820649
You've Got Mail (1998)                 3.380668
Young Frankenstein (1974)              4.250629
Young Guns (1988)                      3.418149

[617 rows x 1 columns]

#### Top rated movie in the given datasets is below
- sorted the obtained values in descending order by using the below function to get the max mean values on top

In [15]:
highest_ratings = ratings_users_movies_mean.sort_values(by='mean', ascending=False)
highest_ratings.head()

mean
title                                                       
Seven Samurai (The Magnificent Seven) (Shichini...  4.560510
Shawshank Redemption, The (1994)                    4.554558
Godfather, The (1972)                               4.524966
Close Shave, A (1995)                               4.520548
Usual Suspects, The (1995)                          4.517106

#### Highest rated movie in the given dataset is Seven Samurai (The Magnificent Seven) (Shichinin no samurai) (1954)	with a mean rating of 4.560510
- Top rated movies are listed above.

###### What is a good rated movie for date night? (60 pts)

- Hint - highly rated movie by 
    - both partners (might be the same gender or not),
    - based on genre preferences,
    - age group can also be combined

- Taking the merged dataset and extracting unique values in the genres column
- https://www.geeksforgeeks.org/get-unique-values-from-a-column-in-pandas-dataframe/

In [16]:
ratings_users_movies.genres.unique()

array(["Animation|Children's|Comedy", "Adventure|Children's|Fantasy",
       'Comedy|Romance', 'Comedy|Drama', 'Comedy',
       'Action|Crime|Thriller', "Adventure|Children's", 'Action',
       'Action|Adventure|Thriller', 'Comedy|Drama|Romance',
       'Comedy|Horror', "Animation|Children's", 'Drama',
       'Action|Adventure|Romance', 'Drama|Thriller', 'Drama|Romance',
       'Thriller', 'Action|Comedy|Drama', 'Crime|Drama|Thriller',
       'Drama|Sci-Fi', 'Romance', 'Adventure|Sci-Fi', 'Adventure|Romance',
       "Children's|Comedy|Drama", 'Documentary', 'Drama|War',
       'Action|Crime|Drama', 'Action|Adventure', 'Crime|Thriller',
       "Animation|Children's|Musical|Romance", "Children's|Comedy",
       'Drama|Mystery', 'Sci-Fi|Thriller',
       'Action|Comedy|Crime|Horror|Thriller', 'Drama|Musical',
       'Crime|Drama|Romance', 'Adventure|Drama', 'Action|Thriller',
       "Adventure|Children's|Comedy|Musical", 'Action|Drama|War',
       'Action|Adventure|Crime', 'Crime', 'Drama

##### to extract the best night movie, I set one of the filter (age).
- I filter the data only for the age group in between 22 and 30

In [17]:
ra_us_mo = ratings_users_movies.loc[(ratings_users_movies['age'] > 22) & (ratings_users_movies['age'] < 30)]

In [18]:
ra_us_mo

movie_id                  title                       genres  year  \
2               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
3               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
9               1       Toy Story (1995)  Animation|Children's|Comedy  1995   
10              1       Toy Story (1995)  Animation|Children's|Comedy  1995   
12              1       Toy Story (1995)  Animation|Children's|Comedy  1995   
...           ...                    ...                          ...   ...   
1000198      3952  Contender, The (2000)               Drama|Thriller  2000   
1000201      3952  Contender, The (2000)               Drama|Thriller  2000   
1000204      3952  Contender, The (2000)               Drama|Thriller  2000   
1000205      3952  Contender, The (2000)               Drama|Thriller  2000   
1000206      3952  Contender, The (2000)               Drama|Thriller  2000   

            short_title  user_id gender  age  occupation    zip  rating  \
2             Toy Story        8      M   25          12  11413       4   
3             Toy Story        9      M   25          17  61614       5   
9             Toy Story       26      M   25           7  23112       3   
10            Toy Story       28      F   25           1  14607       3   
12            Toy Story       36      M   25           3  94123       5   
...                 ...      ...    ...  ...         ...    ...     ...   
1000198  Contender, The     5333      F   25           7  02332       4   
1000201  Contender, The     5475      M   25           4  94110       5   
1000204  Contender, The     5812      F   25           7  92120       4   
1000205  Contender, The     5831      M   25           1  92120       3   
1000206  Contender, The     5837      M   25           7  60607       4   

          timestamp  
2         978233496  
3         978225952  
9         978130703  
10        978985309  
12        978061285  
...             ...  
1000198   984385603  
1000201   994750078  
1000204   992072099  
1000205   986223125  
1000206  1011902656  

[395556 rows x 12 columns]

- selecting the data based on the preferred genres

In [19]:
genres = ra_us_mo.loc[ra_us_mo['genres'].str.contains('Sci-Fi' and 'Drama' and 'Romance' and 'Fantasy')]
genres

# reference :
# https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html

movie_id                                              title  \
2081           2                                     Jumanji (1995)   
2084           2                                     Jumanji (1995)   
2085           2                                     Jumanji (1995)   
2089           2                                     Jumanji (1995)   
2092           2                                     Jumanji (1995)   
...          ...                                                ...   
996111      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996117      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996118      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996120      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996122      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   

                              genres  year  \
2081    Adventure|Children's|Fantasy  1995   
2084    Adventure|Children's|Fantasy  1995   
2085    Adventure|Children's|Fantasy  1995   
2089    Adventure|Children's|Fantasy  1995   
2092    Adventure|Children's|Fantasy  1995   
...                              ...   ...   
996111                 Drama|Fantasy  1993   
996117                 Drama|Fantasy  1993   
996118                 Drama|Fantasy  1993   
996120                 Drama|Fantasy  1993   
996122                 Drama|Fantasy  1993   

                                         short_title  user_id gender  age  \
2081                                         Jumanji       27      M   25   
2084                                         Jumanji       48      M   25   
2085                                         Jumanji       53      M   25   
2089                                         Jumanji       83      F   25   
2092                                         Jumanji      114      F   25   
...                                              ...      ...    ...  ...   
996111  Faraway, So Close (In Weiter Ferne, So Nah!)     3916      M   25   
996117  Faraway, So Close (In Weiter Ferne, So Nah!)     4607      M   25   
996118  Faraway, So Close (In Weiter Ferne, So Nah!)     4950      M   25   
996120  Faraway, So Close (In Weiter Ferne, So Nah!)     5371      M   25   
996122  Faraway, So Close (In Weiter Ferne, So Nah!)     5787      M   25   

        occupation    zip  rating   timestamp  
2081            11  19130       1   978129692  
2084             4  92107       3   978064964  
2085             0  96931       5   977981548  
2089             2  94609       3   977721051  
2092             2  83712       4   977506644  
...            ...    ...     ...         ...  
996111           4  02134       4   969466671  
996117           0  27403       4   998920139  
996118          16  55421       4   975131185  
996120          11  55408       5   978984041  
996122          20  90802       4  1001123685  

[14290 rows x 12 columns]

In [20]:
male_choices = genres.loc[genres['gender'] == 'M']
male_choices
# extracting ratings given by males for the selected genres

movie_id                                              title  \
2081           2                                     Jumanji (1995)   
2084           2                                     Jumanji (1995)   
2085           2                                     Jumanji (1995)   
2099           2                                     Jumanji (1995)   
2102           2                                     Jumanji (1995)   
...          ...                                                ...   
996111      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996117      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996118      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996120      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996122      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   

                              genres  year  \
2081    Adventure|Children's|Fantasy  1995   
2084    Adventure|Children's|Fantasy  1995   
2085    Adventure|Children's|Fantasy  1995   
2099    Adventure|Children's|Fantasy  1995   
2102    Adventure|Children's|Fantasy  1995   
...                              ...   ...   
996111                 Drama|Fantasy  1993   
996117                 Drama|Fantasy  1993   
996118                 Drama|Fantasy  1993   
996120                 Drama|Fantasy  1993   
996122                 Drama|Fantasy  1993   

                                         short_title  user_id gender  age  \
2081                                         Jumanji       27      M   25   
2084                                         Jumanji       48      M   25   
2085                                         Jumanji       53      M   25   
2099                                         Jumanji      149      M   25   
2102                                         Jumanji      195      M   25   
...                                              ...      ...    ...  ...   
996111  Faraway, So Close (In Weiter Ferne, So Nah!)     3916      M   25   
996117  Faraway, So Close (In Weiter Ferne, So Nah!)     4607      M   25   
996118  Faraway, So Close (In Weiter Ferne, So Nah!)     4950      M   25   
996120  Faraway, So Close (In Weiter Ferne, So Nah!)     5371      M   25   
996122  Faraway, So Close (In Weiter Ferne, So Nah!)     5787      M   25   

        occupation    zip  rating   timestamp  
2081            11  19130       1   978129692  
2084             4  92107       3   978064964  
2085             0  96931       5   977981548  
2099             1  29205       3   977591827  
2102            12  10458       3  1015890397  
...            ...    ...     ...         ...  
996111           4  02134       4   969466671  
996117           0  27403       4   998920139  
996118          16  55421       4   975131185  
996120          11  55408       5   978984041  
996122          20  90802       4  1001123685  

[11266 rows x 12 columns]

In [21]:
male_choices1 = male_choices['title'].value_counts()
male_choices1




Star Wars: Episode IV - A New Hope (1977)           899
Star Wars: Episode I - The Phantom Menace (1999)    733
E.T. the Extra-Terrestrial (1982)                   669
Beetlejuice (1988)                                  479
Big (1988)                                          462
                                                   ... 
Star Kid (1997)                                      15
Bogus (1996)                                         14
One Magic Christmas (1985)                            5
Simple Wish, A (1997)                                 4
Kids of the Round Table (1995)                        3
Name: title, Length: 67, dtype: int64

In [22]:
date_night = male_choices[male_choices['title'].isin(male_choices1[male_choices1 > 500].index)]
date_night

movie_id                                             title  \
67448        260         Star Wars: Episode IV - A New Hope (1977)   
67452        260         Star Wars: Episode IV - A New Hope (1977)   
67459        260         Star Wars: Episode IV - A New Hope (1977)   
67460        260         Star Wars: Episode IV - A New Hope (1977)   
67464        260         Star Wars: Episode IV - A New Hope (1977)   
...          ...                                               ...   
707186      2628  Star Wars: Episode I - The Phantom Menace (1999)   
707187      2628  Star Wars: Episode I - The Phantom Menace (1999)   
707188      2628  Star Wars: Episode I - The Phantom Menace (1999)   
707189      2628  Star Wars: Episode I - The Phantom Menace (1999)   
707193      2628  Star Wars: Episode I - The Phantom Menace (1999)   

                                 genres  year  \
67448   Action|Adventure|Fantasy|Sci-Fi  1977   
67452   Action|Adventure|Fantasy|Sci-Fi  1977   
67459   Action|Adventure|Fantasy|Sci-Fi  1977   
67460   Action|Adventure|Fantasy|Sci-Fi  1977   
67464   Action|Adventure|Fantasy|Sci-Fi  1977   
...                                 ...   ...   
707186  Action|Adventure|Fantasy|Sci-Fi  1999   
707187  Action|Adventure|Fantasy|Sci-Fi  1999   
707188  Action|Adventure|Fantasy|Sci-Fi  1999   
707189  Action|Adventure|Fantasy|Sci-Fi  1999   
707193  Action|Adventure|Fantasy|Sci-Fi  1999   

                                      short_title  user_id gender  age  \
67448          Star Wars: Episode IV - A New Hope        3      M   25   
67452          Star Wars: Episode IV - A New Hope       15      M   25   
67459          Star Wars: Episode IV - A New Hope       26      M   25   
67460          Star Wars: Episode IV - A New Hope       27      M   25   
67464          Star Wars: Episode IV - A New Hope       36      M   25   
...                                           ...      ...    ...  ...   
707186  Star Wars: Episode I - The Phantom Menace     6019      M   25   
707187  Star Wars: Episode I - The Phantom Menace     6021      M   25   
707188  Star Wars: Episode I - The Phantom Menace     6023      M   25   
707189  Star Wars: Episode I - The Phantom Menace     6024      M   25   
707193  Star Wars: Episode I - The Phantom Menace     6030      M   25   

        occupation    zip  rating  timestamp  
67448           15  55117       5  978297512  
67452            7  22903       4  978212645  
67459            7  23112       3  978271884  
67460           11  19130       5  978129642  
67464            3  94123       5  993701943  
...            ...    ...     ...        ...  
707186           0  10024       4  956760889  
707187          12  08876       4  956756903  
707188           0  43213       5  956751100  
707189          12  53705       3  956749699  
707193          17  32618       3  956718105  

[2301 rows x 12 columns]

In [23]:
# ratings_users_movies_2 = ratings_users_movies_1.groupby('title')['rating'].agg(['mean'])
date_night_movie = date_night.groupby('title')['rating'].agg(['mean'])

In [24]:
date_night_movie

mean
title                                                     
E.T. the Extra-Terrestrial (1982)                 3.911809
Star Wars: Episode I - The Phantom Menace (1999)  3.398363
Star Wars: Episode IV - A New Hope (1977)         4.614016

In [25]:
date_night_movie = date_night_movie.sort_values(by='mean', ascending=False)
date_night_movie
# this is the output for the best date night movie (when only male ratings are considered)

mean
title                                                     
Star Wars: Episode IV - A New Hope (1977)         4.614016
E.T. the Extra-Terrestrial (1982)                 3.911809
Star Wars: Episode I - The Phantom Menace (1999)  3.398363

In [26]:
female_choices = genres.loc[genres['gender'] == 'F']
female_choices

movie_id                                              title  \
2089           2                                     Jumanji (1995)   
2092           2                                     Jumanji (1995)   
2109           2                                     Jumanji (1995)   
2137           2                                     Jumanji (1995)   
2157           2                                     Jumanji (1995)   
...          ...                                                ...   
991573      3889                         Highlander: Endgame (2000)   
991585      3889                         Highlander: Endgame (2000)   
991630      3889                         Highlander: Endgame (2000)   
996059      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   
996075      3920  Faraway, So Close (In Weiter Ferne, So Nah!) (...   

                              genres  year  \
2089    Adventure|Children's|Fantasy  1995   
2092    Adventure|Children's|Fantasy  1995   
2109    Adventure|Children's|Fantasy  1995   
2137    Adventure|Children's|Fantasy  1995   
2157    Adventure|Children's|Fantasy  1995   
...                              ...   ...   
991573      Action|Adventure|Fantasy  2000   
991585      Action|Adventure|Fantasy  2000   
991630      Action|Adventure|Fantasy  2000   
996059                 Drama|Fantasy  1993   
996075                 Drama|Fantasy  1993   

                                         short_title  user_id gender  age  \
2089                                         Jumanji       83      F   25   
2092                                         Jumanji      114      F   25   
2109                                         Jumanji      262      F   25   
2137                                         Jumanji      475      F   25   
2157                                         Jumanji      692      F   25   
...                                              ...      ...    ...  ...   
991573                           Highlander: Endgame     1812      F   25   
991585                           Highlander: Endgame     2192      F   25   
991630                           Highlander: Endgame     5996      F   25   
996059  Faraway, So Close (In Weiter Ferne, So Nah!)      507      F   25   
996075  Faraway, So Close (In Weiter Ferne, So Nah!)     1394      F   25   

        occupation    zip  rating  timestamp  
2089             2  94609       3  977721051  
2092             2  83712       4  977506644  
2109             1  68503       3  976661088  
2137             2  55421       4  976219986  
2157             0  85207       5  975739161  
...            ...    ...     ...        ...  
991573          12  48103       4  974854457  
991585           0  27713       1  974606920  
991630           0  87114       5  969386663  
996059           0  55405       5  976207639  
996075           3  94117       4  974764308  

[3024 rows x 12 columns]

In [27]:
female_choices1 = female_choices['title'].value_counts()
female_choices1

Star Wars: Episode IV - A New Hope (1977)              229
E.T. the Extra-Terrestrial (1982)                      217
Star Wars: Episode I - The Phantom Menace (1999)       170
Beetlejuice (1988)                                     150
Big (1988)                                             146
                                                      ... 
One Magic Christmas (1985)                               2
Faraway, So Close (In Weiter Ferne, So Nah!) (1993)      2
Kazaam (1996)                                            1
Warriors of Virtue (1997)                                1
Star Kid (1997)                                          1
Name: title, Length: 66, dtype: int64

In [28]:
date_night1 = female_choices[female_choices['title'].isin(female_choices1[female_choices1 > 100].index)]
date_night1

movie_id                                      title  \
67457        260  Star Wars: Episode IV - A New Hope (1977)   
67461        260  Star Wars: Episode IV - A New Hope (1977)   
67476        260  Star Wars: Episode IV - A New Hope (1977)   
67477        260  Star Wars: Episode IV - A New Hope (1977)   
67483        260  Star Wars: Episode IV - A New Hope (1977)   
...          ...                                        ...   
757843      2797                                 Big (1988)   
757852      2797                                 Big (1988)   
757860      2797                                 Big (1988)   
757863      2797                                 Big (1988)   
757871      2797                                 Big (1988)   

                                 genres  year  \
67457   Action|Adventure|Fantasy|Sci-Fi  1977   
67461   Action|Adventure|Fantasy|Sci-Fi  1977   
67476   Action|Adventure|Fantasy|Sci-Fi  1977   
67477   Action|Adventure|Fantasy|Sci-Fi  1977   
67483   Action|Adventure|Fantasy|Sci-Fi  1977   
...                                 ...   ...   
757843                   Comedy|Fantasy  1988   
757852                   Comedy|Fantasy  1988   
757860                   Comedy|Fantasy  1988   
757863                   Comedy|Fantasy  1988   
757871                   Comedy|Fantasy  1988   

                               short_title  user_id gender  age  occupation  \
67457   Star Wars: Episode IV - A New Hope       24      F   25           7   
67461   Star Wars: Episode IV - A New Hope       28      F   25           1   
67476   Star Wars: Episode IV - A New Hope       81      F   25           0   
67477   Star Wars: Episode IV - A New Hope       83      F   25           2   
67483   Star Wars: Episode IV - A New Hope       96      F   25          16   
...                                    ...      ...    ...  ...         ...   
757843                                 Big     5925      F   25           0   
757852                                 Big     5955      F   25           7   
757860                                 Big     5972      F   25          20   
757863                                 Big     5996      F   25           0   
757871                                 Big     6036      F   25          15   

               zip  rating  timestamp  
67457        10023       5  978135156  
67461        14607       4  978125450  
67476        60640       5  977786694  
67477        94609       5  977720825  
67483        78028       5  980562704  
...            ...     ...        ...  
757843  90035-4444       4  957294321  
757852       34952       4  957108406  
757860       55428       4  963614410  
757863       87114       5  956854231  
757871       32603       3  956717085  

[1278 rows x 12 columns]

In [29]:
date_night_movie1 = date_night1.groupby('title')['rating'].agg(['mean'])
date_night_movie1

mean
title                                                     
Beetlejuice (1988)                                3.686667
Big (1988)                                        3.938356
E.T. the Extra-Terrestrial (1982)                 4.096774
Mask, The (1994)                                  3.107843
Splash (1984)                                     3.546154
Star Wars: Episode I - The Phantom Menace (1999)  3.258824
Star Wars: Episode IV - A New Hope (1977)         4.410480
Willy Wonka and the Chocolate Factory (1971)      4.208955

In [30]:
date_night_movie1 = date_night_movie1.sort_values(by='mean', ascending=False)
date_night_movie1
# this is the output for the best date night movie (when only female ratings are considered)

mean
title                                                     
Star Wars: Episode IV - A New Hope (1977)         4.410480
Willy Wonka and the Chocolate Factory (1971)      4.208955
E.T. the Extra-Terrestrial (1982)                 4.096774
Big (1988)                                        3.938356
Beetlejuice (1988)                                3.686667
Splash (1984)                                     3.546154
Star Wars: Episode I - The Phantom Menace (1999)  3.258824
Mask, The (1994)                                  3.107843

- Based on both the partners preferred genres, I extracted the top rated movie [for the specific genres only] seperately for the male preference and female preference


### good rated movie for date night is Star Wars: Episode IV - A New Hope (1977)
    - Based on top ratings and for preferred genres of the partners
    - Based on both the gender preferences
    - for specific age group in between 22 and 30